Update pull_request_template.md

lesliecdubs opened the pull request:
Adding more depth to the standard PRC PR template to ensure library maintainers and contributors have provided enough context in order to confidently review changes before merge.

broccolinisoup reviewed (approved):
Looks great! Thank you for this PR.

lesliecdubs commented:
Thanks for the ideas @joshblack. I'd love for the changes we decide on for these templates to be instated across both PRC and PVC to keep the contribution experience more seamless between libraries, and since the content is generic enough to be relevant across both. That said, I'm cc'ing the PVC folks here in case they would like to weigh in too: @primer/rails-reviewers

My initial thoughts:

> Have Closes # at the top of the issue

I think this is analogous to the new "List the issues that this change affects" heading. I moved that up to the top in [this commit](https://github.com/primer/react/pull/3344/commits/9053cf2f11ead572034dc4293a1577bfa7cb0419), but feel free to suggest further changes directly on this PR.

> Consolidate certain sections into a general "about" section, specifically:
> What are you trying to accomplish
> What approach did you choose and why

This is okay with me assuming it doesn't mean we'll get paragraphs of info in that general section. @joshblack can you drop a suggestion into this PR with the change you're proposing?

> Remove Screenshots as a top-level heading and roll it into About above

@joshblack can you add a suggestion for this change, and explicitly mention "screenshots" in instructions in the "about" section? I think it's fine for PRs to go up without a screenshot if they don't significantly impact UI, but I would still expect screenshots to be added for any relevant changes.

> Add an explicit Changelog section that details the changes the PR makes from the perspective of an end user, e.g. what has been added, changed, or removed

Is this for the benefit of reuse in our release notes, or for the benefit of the reviewer? Either way, seems like a reasonable addition to me.

> Add a "Testing & Reviewing" section
> This provides a way to provide reviewers with specific context or things to pay attention to when reviewing
> This could also include Merge checklist
> Remove the Accessibility section, I'm not really sure how to apply this to the work that we have and what it would add given that this work is part of our practices

I'm not too precious about the current heading structure in the template, so if you'd like to propose changes please feel free. That said, I think it's important that we maintain:
-  a merge checklist of key actions that must be done/true before a reviewer will merge
- a callout for accessibility, whether this is that automatic axe checks should pass in CI or that any UI work has been tested manually with a screenreader and/or Accessibility Insights tools

joshblack commented on .github/pull_request_template.md:
```diff
@@ -1,15 +1,52 @@
+### What are you trying to accomplish?
+<!-- Provide a description of the changes. -->
 
 ### Screenshots
+<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->
```
```suggestion
```

lesliecdubs replied:
Can you provide any more context for this suggestion? The instructions note that this section is fine to remove if the changes do not impact UI.

lesliecdubs replied:
I see now that you moved the screenshot suggestion to elsewhere in the template. That's fine with me; I'll push a recommended change to the instruction there but otherwise I'm good with this.

joshblack commented on .github/pull_request_template.md:
```diff
@@ -1,15 +1,52 @@
-Describe your changes here.
+_Authors: Please fill out this form carefully and completely._
 
-Closes # (type the issue number after # if applicable; otherwise remove this line)
+_Reviewers: By approving this Pull Request you are approving the code change, as well as its deployment and mitigation plans._
+_Please read this description carefully. If you feel there is anything unclear or missing, please ask for updates._
+
+#### List the issues that this change affects.
+<!--Every code change must address _at least 1_ issue. Fixes a bug, completes a task, every change
+      should have a corresponding issue listed here. If one does not already exist, create one. -->
+      
+Closes # (type the GitHub issue number after #)
+
+### What are you trying to accomplish?
+<!-- Provide a description of the changes. -->
```
```suggestion
<!-- Provide the GitHub issue that this issue closes. Start typing the number or name of the issue after the # below -->

Closes #

<!-- Provide an overview of the changes, including any screenshots when helpful -->
```

lesliecdubs replied:
Resolved in commit https://github.com/primer/react/pull/3344/commits/6631c1b2ea7cfe9226201b606ef0261ba57ab1b8

joshblack commented on .github/pull_request_template.md:
```diff
@@ -1,15 +1,52 @@
+     Identify any work you did to mitigate risk.
+     Describe any alternative approaches you considered and why you discarded them. -->
+
+### Anything you want to highlight for special attention from reviewers?
+<!-- This is your chance to identify remaining risks and confess any uncertainties you may have about the correctness of the changes.
+     Highlight anything on which you would like a second (or third) opinion.
+     Keep in mind how many component uses cases may be affected by your changes when assessing risk. -->
```
I think this section could be framed as "testing & reviewing" in order to provide a way for authors to provide reviewers with specific context into the PR and how it could be tested/reviewed

```suggestion
## Testing & Reviewing

<!-- Describe any specific details to help reviewers test or review this Pull Request -->
```

lesliecdubs replied:
Updated, see diff: https://github.com/primer/react/pull/3344/files#diff-b2496e80299b8c3150b1944450bd81c622e04e13d15c411d291db0927d75fd6bR21

joshblack commented on .github/pull_request_template.md:
```diff
@@ -1,15 +1,52 @@
 ### Screenshots
+<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->
+
+#### Risk Assessment
+  <!-- Please select from one of the following and detail why this level was chosen -->
+
+- [ ] **Low risk** the change is small, highly observable, and easily rolled back.
+- [ ] **Medium risk** changes that are isolated, reduced in scope or could impact few users. The change will not impact library availability.
+- [ ] **High risk** changes are those that could impact customers and SLOs, low or no test coverage, low observability, or slow to rollback.
```
I think this section is challenging from the library side as it's hard or limiting to express changes in terms of:

- Observability
- Availability
- The ability of a change to be rolled back
- The impact of a change to SLOs

Instead of evaluating the risk of a change as it relates to a production environment, we could talk about the types of changes that the PR introduces from the perspective of the end user.

```suggestion
## Changelog

<!-- List out API changes that this Pull Request introduces -->

### New

### Changed

### Removed
```

lesliecdubs replied:
Adopted the recommended change and added a section for "rollout strategy": https://github.com/primer/react/pull/3344/commits/62bedc473eff54e11c46cd0a4794e0d9dcb7d310

joshblack commented on .github/pull_request_template.md:
```diff
@@ -1,15 +1,52 @@
+- [ ] **Medium risk** changes that are isolated, reduced in scope or could impact few users. The change will not impact library availability.
+- [ ] **High risk** changes are those that could impact customers and SLOs, low or no test coverage, low observability, or slow to rollback.
+
+### What approach did you choose and why?
+<!-- This section is a place for you to describe your thought process in making these changes.
+     List any tradeoffs you made to take on or pay down tech debt.
+     Identify any work you did to mitigate risk.
+     Describe any alternative approaches you considered and why you discarded them. -->
```
This seems like something that can be rolled into the overview section above

```suggestion
```

lesliecdubs replied:
Removed this section in https://github.com/primer/react/pull/3344/commits/8fb13a10eb835d35835123fbcefa12a572c96e8c

joshblack commented on .github/pull_request_template.md:
```diff
@@ -1,15 +1,52 @@
+     Keep in mind how many component uses cases may be affected by your changes when assessing risk. -->
 
-Please provide before/after screenshots for any visual changes
+### Accessibility
+<!--
+  You may remove this section and the "Accessibility" heading above _only_ if the changes in this pull request do not impact UI. Delete all those that don't apply.
+  If there are any accessibility-related updates, please describe them here.
+-->
+* **Fixes axe scan violation** - This change fixes an existing [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violation.
+* **No new axe scan violation** - This change does not introduce any new [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violations.
+* **New axe violation** - This change introduces a new [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violation. Please describe why the violation cannot be resolved below.
```
This section feels like a "practice" detail in that accessibility is a part of the work we do and that work must not introduce a11y/axe violations.

For example, in what scenario would this section be filled out with "New axe violation" selected as an option? It seems like this would mostly be "No new axe scan violations" and occasionally "Fixes axe scan violations" but this detail could be included in the changelog section above

lesliecdubs replied:
Are we running axe scans in CI in `primer/react` yet, or is that still tbd? If the answer is yes, I think we can remove this whole section. 

If not, I'd probably prefer to maintain a section dedicated to accessibility for now to prompt contributors to ensure compliance when opening the PR.

joshblack replied:
> Are we running axe scans in CI in primer/react yet, or is that still tbd? If the answer is yes, I think we can remove this whole section.

@lesliecdubs it should be enabled and required at this point 👍

lesliecdubs replied:
Thanks for confirming. I've removed this redundant section of the template in https://github.com/primer/react/pull/3344/commits/48815b5dd6bef7ab2b565183995c4cf0d9d8dea1.

joshblack reviewed (commented):
Tried to leave review comments, hope that they make sense 🤔 This kind of context can make it challenging to request changes as they might not make sense in isolation.

lesliecdubs commented:
Thanks for the feedback @joshblack! I iterated a bit more based on your thoughts and subsequent conversations. Would love your 👀s again! 

Also added a review request for other key PRC maintainers before we merge.

joshblack commented on .github/pull_request_template.md:
```diff
@@ -1,15 +1,42 @@
-Describe your changes here.
+### What are you trying to accomplish?
```
```suggestion
```

joshblack commented on .github/pull_request_template.md:
```diff
@@ -1,15 +1,42 @@
+<!-- Provide an overview of the changes, including before/after screenshots, videos, or graphs when helpful -->
 
-### Screenshots
+#### Changelog & rollout strategy
```
```suggestion
### Changelog
```

joshblack commented on .github/pull_request_template.md:
```diff
@@ -1,15 +1,42 @@
-### Screenshots
+#### Changelog & rollout strategy
+<!-- Under the headings below, list out relevant API changes that this Pull Request introduces -->
+### New
```
Updated heading level based on above change and added comment for section

```suggestion
#### New

<!-- List of things added in this PR -->

#### Changed

<!-- List of things changed in this PR -->

#### Removed

<!-- List of things removed in this PR -->
```

joshblack reviewed (commented):
Left a couple suggestions for hierarchy and a couple comments that could be useful, thanks for putting this all together!

joshblack commented:
> Also not sure how to clear this [failing status check](https://github.com/primer/react/actions/runs/6125661359/job/16628613525?pr=3344#step:5:19); any tips?

@lesliecdubs I think running `npm run format` should do the trick! Just needs to be formatted by prettier
